dips/dip-158.md

```diff
@@ -0,0 +1,805 @@
+|national_id|[NationalIdObject](#nationalidobject)|N|National ID information for payer. e.g. SSN|
+
+```
+"payer_data": 
+{
+  "name": "Alice",
+  "surname": "Cooper",
+  "address": {
+    AddressObject()
+  },
+  "national_id": {
+    NationalIdObject()
+  }        
+}
+``` 
+## BusinessDataObject
+The details of the merchant.
+|Field|Type|Required?|Description|
+|-|-|-|-|
+|name|str|Y|Merchant's display name. Should be recognizable by the payer|
+|legal_name|str|Y|The legal entity name|
+|image_url|str|N|URL with the business logo|
+|address|[AddressObject](#addressobject)|Y|Merchant's physical address information|
+```
+"business_data": 
+{
+  "name": "Acme",
+  "legal_name": "Acme Technologies Inc.",
+  "address": {
+    AddressObject()
+  }        
+}
+``` 
+
+## PaymentActionObject
+Information about the payment type (i.e. action), amount and currency.
+
+|Field|Type|Required?|Description|
+|-----|----|---------|-----------|
+| amount | uint | Y | Amount of the transfer.  Base units are the same as for on-chain transactions for this currency.  For example, if DiemUSD is represented on-chain where "1" equals 1e-6 dollars, then "1" equals the same amount here.  For any currency, the on-chain mapping must be used for amounts. |
+| currency | enum | Y | One of the supported on-chain currency types, e.g. XUS|
+| action | [Action enum](#action-enum) | Y | Populated in the request.  This value indicates the requested action to perform. For immediate capture and P2P transfer, "charge" is used.  For auth and capture, "auth" and "capture" should be used. "capture" can only be performed after a valid "auth" |
+| valid_until | uint | N | Unix timestamp indicating until when the payment can be authorized. Once this time has been reached, an attempt to authorize and lock the funds will be rejected. One should consider limiting this period to several minutes (e.g. 30 minutes) which should sufficve for the customer to approve or reject the payment request|
+| timestamp | uint | Y | Unix timestamp indicating the time that the payment Command was created.|
+
+```
+{
+    "amount": 100000000,
+    "currency": "XUS",
+    "action": "auth",
+    "valid_until": 74000,
+    "timestamp": 72322,
+}
+```
+### Action Enum
+The following values are allowed for the Action Enum
+* `charge`
+* `auth`
+* `capture`
+
+## PaymentStatusObject
+| Field | Type | Required? | Description |
+|-|-|-|-|
+| status | [Status enum](#status-enum) | Y | Status of the payment|
+| abort_code |[AbortCode enum](#abortcode-enum)| N | In the case of an `abort` status, this field may be used to describe the reason for the abort. Represents the error code of the corresponding error|
+| abort_message | str | N | Additional details about this error.  To be used only when `code` is populated |
+```
+"status": {
+  "status": "abort",
+  "abort_code": "void",
+  "abort_message": "abort message"
+}
+```
+
+### Status Enum
+* `none` - Initial state
```

Comment:
I think there's confusion, this status is actually a response status and not necessarily a state transition. I think the two have been a bit conflated in the following and would recommend we eliminate merging the two concepts.

For example, none and needs_init_data never hit the wire. I could see arguments for the other 3 unless we wanted to introduce error responses instead of status objects, which I kinda wouldn't mind since I like exceptions.

Reply:
I like the idea of response codes instead of statuses.
Since each command type indicates a specific action/operation, the command response can indicate success or failure w/o specifying the status. If the command response is "failure" it will include also an error code and error message.
I will replace the PaymentStatusObject with a new structure PaymentCommandResultObject.
It will include the following fields:
1. result - enum with two possible values: success, error
2. error_code - enum with relevant error codes. for example: risk_checks_failed, payment_type_mismatch. Relevant only if result = error
3. error_message - a string containing additional details about the error (optional). Relevant only if result = error

This will significantly reduce the importance of mentioning the different statuses and allowed transitions in the DIP. In fact, we should consider removing them. Let's discuss.